graph/patterns/upsert.md

```diff
@@ -0,0 +1,209 @@
+* NOTE: the service must also support `GET` using the alternate key pattern.
+* For consistent CRUD Microsoft Graph behaviors, all resources, **including** resources used in IaC scenarios, should use `POST` and a service-generated primary key, per existing guidelines, and support `GET`, `PATCH` and `DELETE` using the primary key.
+* If a service does not support `Upsert`, then a `PATCH` call against a non-existent resource must result in an HTTP "409 conflict" error.
+
+This solution allows for existing resources that follow Microsoft Graph conventions for CRUD operations to add `Upsert` without impacting existing apps or functionality.
+
+Ideally, all new entity types should support an `Upsert` mechanism, especially where they support control-plane APIs, or are used in admin style or IaC scenarios.
+
+## When to use this pattern
+
+This pattern should be adopted for resources that are managed through infrastructure as code or desired state configuration.
+
+## Issues and considerations
+
+* The addition of this new pattern (with alternate key) does not represent a breaking change.
+However, some API producers may have concerns about accidental usages of this new pattern unwittingly creating many new resources when the intent was an update.
+As a result, API producers can use the `Prefer: idempotent` to require clients to opt-in to the Upsert behavior.
+* The client-provided alternate key must be immutable after being set. If its value is null then it should be settable as a way to backfill existing resources for use in IaC scenarios.
+* API producers could use `PUT` operations to create or update, but generally this approach is not recommended due to the destructive nature of `PUT`'s replace semantics.
+* API producers could use `Upsert` with a primary (client-provided) key and this may be appropriate for some scenarios. However, the recommendation is for resources to support creation using `POST` and a service-generated primary key, for consistency reasons.
+* API producers may annotate entity sets, singletons and collections to indicate that entities can be "upserted". The example below shows this annotation for the `groups` entity set.  
+
+```xml
+<EntitySet Name="groups" EntityType="microsoft.graph.group">
+  <Annotation Term="Org.OData.Capabilities.V1.UpdateRestrictions">
+    <Record>
+      <PropertyValue Property="Upsertable" Bool="true"/>
+    </Record>
+  </Annotation>
+</EntitySet>
+```
+
+## Examples
+
+For these examples we'll use the `group` entity type, which defines both a primary (service-generated) key (`id`) and an alternate (client-provided) key (`uniqueName`).  
+
+```xml
+<EntityType Name="group">
+  <Key>
+    <PropertyRef Name="id"/>
+  </Key>
+  <Property Name="id" Type="Edm.String"/> 
+  <Property Name="uniqueName" Type="Edm.String"/>
+  <Property Name="displayName" Type="Edm.String"/>
+  <Property Name="description" Type="Edm.String"/> 
+  <Annotation Term="OData.Community.Keys.V1.AlternateKeys">
```

Comment:
>OData.Community.Keys.V1. [](http://example.com/codeflow?start=20&length=24)

can we move away from using OData.Community.Keys.V1, since this is now in Core? ([https://github.com/oasis-tcs/odata-vocabularies/blob/main/vocabularies/Org.OData.Core.V1.xml#L479](https://github.com/oasis-tcs/odata-vocabularies/blob/main/vocabularies/Org.OData.Core.V1.xml#L479))?

Reply:
Argh - not sure how that snuck in.  Good catch